src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -2572,9 +2576,116 @@ protected override void HandleReadMemoryRequestAsync(IRequestResponder<ReadMemor
+                }
+            }
 
-#region IDebugPortNotify2
+            foreach (var instructionBp in responder.Arguments.Breakpoints)
+            {
+                if (GetMemoryContext(instructionBp.InstructionReference, instructionBp.Offset, out IDebugMemoryContext2 memoryContext, out _) == HRConstants.E_NOTIMPL)
+                {
+                    responder.SetError(new ProtocolException("Debug Engine does not support InstructionBreakpoints."));
+                    return;
+                }
+
+                if (m_instructionBreakpoints.ContainsKey(instructionBp.InstructionReference))
+                {
+                    IDebugBreakpointRequest2 breakpointRequest;
+                    if (m_instructionBreakpoints[instructionBp.InstructionReference].GetBreakpointRequest(out breakpointRequest) == 0 &&
+                                breakpointRequest is AD7BreakPointRequest ad7BPRequest)
+                    {
+                        // Check to see if this breakpoint has a condition that has changed.
+                        if (!StringComparer.Ordinal.Equals(ad7BPRequest.Condition, instructionBp.Condition))
+                        {
+                            // Condition has been modified. Delete breakpoint so it will be recreated with the updated condition.
+                            var toRemove = m_instructionBreakpoints[instructionBp.InstructionReference];
+                            toRemove.Delete();
+                            m_instructionBreakpoints.Remove(instructionBp.InstructionReference);
+                        }
+                        else
+                        {
+                            if (ad7BPRequest.BindResult != null)
+                            {
+                                response.Breakpoints.Add(ad7BPRequest.BindResult);
+                            }
+                            else
+                            {
+                                response.Breakpoints.Add(new Breakpoint()
+                                {
+                                    Id = (int)ad7BPRequest.Id,
+                                    Verified = true,
+                                    Line = 0
+                                });
+
+                            }
+                            continue;
+                        }
+                    }
+                }
+                else
+                {
+                    IDebugPendingBreakpoint2 pendingBp;
+                    AD7BreakPointRequest pBPRequest = new AD7BreakPointRequest(memoryContext);
+                    int hr = HRConstants.E_FAIL;
+
+                    if (m_engine.CreatePendingBreakpoint(pBPRequest, out pendingBp) == HRConstants.S_OK && pendingBp != null)
+                    {
+                        hr = pendingBp.Bind();
+                    }
+
+                    m_instructionBreakpoints.Add(instructionBp.InstructionReference, pendingBp);
```

Comment:
>m_instructionBreakpoints [](http://example.com/codeflow?start=20&length=24)

Should we still do this if Bind fails?

Reply:
I need to double check what function breakpoints is doing. It seems to do the same thing adding to m_functionBreakpoints, but then re-assigning it to `newBreakpoints` so it seems that `Add` is a no-op.